liby2util-r/src/y2log.cc

```diff
@@ -436,6 +438,32 @@ static string old (const string & filename, int i, const char * suffix) {
     return filename + "-" + numbuf + suffix;
 }
 
+/**
+ * Locking wrapper around shift_log_files_if_needed() to avoid race conditions
+ * when several processes access the log in parallel. This ensures only one
+ * process does the shift.
+ */
+static void shift_log_files_if_needed_locked(string filename)
+{
+    // locking needs a fd, so just open the file
+    int fd = open(filename.c_str(), O_RDONLY);
+
+    // file does not exists or other error, do not shift the files
+    if (fd == -1)
+    return;
+
+    // lock the file exclusively, this blocks until the lock is available
+    if (flock(fd, LOCK_EX) == 0)
```

Comment:
well, my question is how will look resulted rotated files? I expect it will have some small entries, as after lock is released it contains just small content meanwhile written another processes. So is possible to simply skip log rotation or check again that log have reasonable size after getting lock?

Reply:
Check that the log has reasonable size is done in the called method `shift_log_files_if_needed` right at the beginning, so if the process was blocked then after unblocking it should quickly find that the rotation is not needed.

Reply:
ah, then it makes sense. BTW: do you measure how often is this method called? I just worry it can slow down openqa with Y2DEBUG enabled.

Reply:
Running `Y2MAXLOGSIZE=1000000000 rake test:unit` in `yast2-storage-ng` produces 158MB log file with 4.1M lines in 25 seconds. That means it should be pretty quick.

I have tested it with sequential tests (to have better reproducible results and be more close with openQA which also runs only one YaST instance).

The times with the original `yast2-core` were: 3:02 and 3:09 (two runs)
With updated package with enabled locking: 3:05 and 3:05 (two runs)

That means even for huge log (4.1M lines) the slow down is hard to measure. It much more depends on the other factors.

With this simple benchmark
```ruby
require "benchmark"
require "yast"
include Yast::Logger
elapsed = Benchmark.realtime do
  (1..1_000_000).each { |i| log.info("#{i}") }
end
puts "Took: #{elapsed} seconds"
```

I got these results:
Original logging: ~14.3s
With locking enabled: ~16.6s

So for logging 1 milion lines the slowdown is ~2 seconds (~15%). And this is an extreme case which does *only* logging, nothing else.

If you think it is still too much we could enable the locking only when e.g. `PARALLEL_TESTS` environment variable is set to cover only this specific case and leave the original behavior for the other cases.